jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+     * @see #setMaxPriority(int)
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    private void processAndComplete(Request request, Response response, Callback callback)
+    {
+        try
+        {
+            boolean handled = process(request, response, callback);
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} handled={} {}", this, handled, request);
+            if (!handled)
+                Response.writeError(request, response, callback, HttpStatus.NOT_FOUND_404);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} failed {}", this, request, x);
+            Response.writeError(request, response, callback, x);
+        }
+    }
+
+    private void resume()
+    {
+        semaphore.release();
+        ListIterator<Queue<Entry>> iterator = queues.listIterator(queues.size());
+        while (iterator.hasPrevious())
+        {
+            Queue<Entry> queue = iterator.previous();
+            Entry entry = queue.poll();
+            if (entry != null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} resuming {}", this, entry.request);
+                // Try again to acquire a semaphore pass.
+                // If it cannot be acquired, the request may
+                // be re-prioritized with a higher priority.
+                processAndComplete(entry.request, entry.response, entry.callback);
+                return;
+            }
+        }
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} no suspended requests to resume", this);
+    }
+
+    private class Entry implements CyclicTimeouts.Expirable
+    {
+        private final Request request;
+        private final Response response;
+        private final Callback callback;
+        private final int priority;
+        private final long expireNanoTime;
+
+        private Entry(Request request, Response response, Callback callback, int priority)
+        {
+            this.request = request;
+            this.response = response;
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : NanoTime.now() + maxSuspend.toNanos();
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return expireNanoTime;
+        }
+
+        private void expire()
+        {
+            // The request timed out, therefore it never acquired a semaphore pass.
+            boolean removed = queues.get(priority).remove(this);
+            if (removed)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} timeout {}", QoSHandler.this, request);
+                Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
```

Comment:
We need to protect error page handling from this mechanism so that they don't need a semaphore

Reply:
If you can't get a semaphore, then you can writeError, which may be an error page dispatch that will try to get the semaphore again which is not available.